_posts/2023-02-28-kani-internship-projects-2022-stubbing.md

```diff
@@ -0,0 +1,361 @@
+
+Let's see another example where we compute the [factorial](https://en.wikipedia.org/wiki/Factorial) of a number `n`.
+Note that the `factorial` function uses the method `checked_mul` to prevent overflow errors.
+We're interested in verifying the correctness of the `factorial` calculation.
+
+```rust
+fn factorial(n: u64) -> Option<u64> {
+    if n == 0 {
+        return Some(1);
+    }
+    n.checked_mul(factorial(n - 1).unwrap())
+}
+
+#[cfg(kani)]
+#[kani::proof]
+fn verify_factorial() {
+    let n = kani::any();
+    kani::assume(n > 0);
+    if let Some(value) = factorial(n) {
+        assert_eq!(value, n * factorial(n - 1).unwrap());
+    }
+}
+```
+
+If we run this example through Kani, we'll see the following output:
+
+```
+Unwinding recursion factorial iteration 1
+Unwinding recursion factorial iteration 2
+Unwinding recursion factorial iteration 3
+Unwinding recursion factorial iteration 4
+Unwinding recursion factorial iteration 5
+Unwinding recursion factorial iteration 6
+Unwinding recursion factorial iteration 7
+Unwinding recursion factorial iteration 8
+Unwinding recursion factorial iteration 9
+Unwinding recursion factorial iteration 10
+# -- snip --
+```
+
+In this case, it doesn't look like Kani will terminate anytime soon due to the recursive definition of the `factorial` function.
+The code under verification shows bad performance because the underlying engine attempts to unwind all possible recursive calls.
+
+But we can avoid the recursion by pre-computing the values that fit into a `u64` value and replacing `factorial` with another function that gets those values.
+
+```rust
+const FACT: [u64; 21] = [
+    1,
+    1,
+    2,
+    6,
+    24,
+    120,
+    720,
+    5040,
+    40320,
+    362880,
+    3628800,
+    39916800,
+    479001600,
+    6227020800,
+    87178291200,
+    1307674368000,
+    20922789888000,
+    355687428096000,
+    6402373705728000,
+    121645100408832000,
+    2432902008176640000,
+];
+
+fn factorial(n: u64) -> Option<u64> {
+    if n == 0 {
+        return Some(1);
+    }
+    n.checked_mul(factorial(n - 1).unwrap())
+}
+
+#[cfg(kani)]
+fn stub_factorial(n_64: u64) -> Option<u64> {
+    let n = n_64 as usize;
+    if n < FACT.len() {
+        Some(FACT[n])
+    } else {
+        None
+    }
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub(factorial, stub_factorial)]
+fn verify_factorial() {
+    let n = kani::any();
+    kani::assume(n > 0);
+    if let Some(value) = factorial(n) {
+        assert_eq!(value, n * factorial(n - 1).unwrap());
+    }
+}
+```
+
+Let's run this harness with Kani!
+
+```bash
+cargo kani --enable-unstable --enable-stubbing --harness verify_factorial
+```
+
+```
+SUMMARY:
+ ** 0 of 7 failed
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.22135049s
+```
+
+Now Kani verifies the example successfully, and in less than 1 second!
+This time, Kani is able to complete the verification because our stub avoids recursion altogether.
+
+## Risks of Stubbing
```

Comment:
This is a good example, but given the risks involve I'd suggest 2 examples. A common error I saw with people applying stubbing to C code was stubs that didn't capture the side-effects of a function. Is there room to add an example of a function that takes a reference and returns a value? The stub should return an arbitrary value and arbitrarily update the reference.

Reply:
Rust avoids that problem by design: First, it won't allow you to mutate the reference unless you declare the parameter as mutable. If you pass a mutable reference, you cannot have other references to that value. Because of that, the only option to recover the value is to return it, but then the updates to the reference cannot be considered a side effect.

This is much better explained [here](https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html#mutable-references).